src/main/java/org/team1540/robot2024/subsystems/drive/Drivetrain.java

```diff
@@ -60,44 +73,35 @@ public Drivetrain(
                 (targetPose) -> Logger.recordOutput("Odometry/TrajectorySetpoint", targetPose));
     }
 
+    @Override
     public void periodic() {
         gyroIO.updateInputs(gyroInputs);
         Logger.processInputs("Drivetrain/Gyro", gyroInputs);
         for (Module module : modules) {
             module.periodic();
         }
 
-        // Stop moving when disabled
-        if (DriverStation.isDisabled()) {
-            for (Module module : modules) {
-                module.stop();
-            }
-        }
-        // Log empty setpoint states when disabled
         if (DriverStation.isDisabled()) {
+            // Stop moving when disabled
+            for (Module module : modules) module.stop();
+
+            // Log empty setpoint states when disabled
             Logger.recordOutput("SwerveStates/Setpoints", new SwerveModuleState[]{});
             Logger.recordOutput("SwerveStates/SetpointsOptimized", new SwerveModuleState[]{});
         }
 
-        // Update odometry
+        // Calculate module deltas
         SwerveModulePosition[] wheelDeltas = new SwerveModulePosition[4];
         for (int i = 0; i < 4; i++) {
             wheelDeltas[i] = modules[i].getPositionDelta();
         }
-        // The twist represents the motion of the robot since the last
-        // loop cycle in x, y, and theta based only on the modules,
-        // without the gyro. The gyro is always disconnected in simulation.
-        Twist2d twist = kinematics.toTwist2d(wheelDeltas);
-        if (gyroInputs.connected) {
-            // If the gyro is connected, replace the theta component of the twist
-            // with the change in angle since the last loop cycle.
-            twist =
-                    new Twist2d(
-                            twist.dx, twist.dy, gyroInputs.yawPosition.minus(lastGyroRotation).getRadians());
-            lastGyroRotation = gyroInputs.yawPosition;
-        }
-        // Apply the twist (change since last loop cycle) to the current pose
-        pose = pose.exp(twist);
+        // Use gyro rotation if gyro is connected, otherwise use module deltas to calculate rotation delta
+        rawGyroRotation =
+                gyroInputs.connected ?
+                        gyroInputs.yawPosition
+                        : rawGyroRotation.plus(Rotation2d.fromRadians(kinematics.toTwist2d(wheelDeltas).dtheta));
```

Comment:
Did you make this up or steal it?

Reply:
Part of this is made up, part of this is stolen. This is so the `SwerveDrivePoseEstimator` can work without a gyro.